### PR TITLE
Update [Resource]Data to inherit the exact matching base class

### PR DESCRIPTION
**Pull request by YalinLi0312**
[ADO item](https://dev.azure.com/azure-mgmt-ex/DotNET%20Management%20SDK/_boards/board/t/DotNET%20Management%20SDK%20Team/Stories/?workitem=5384)

### CODE REVIEW

File: src/AutoRest.CSharp/Output/Models/Types/OutputLibrary.cs [hunk 1]
```diff
@@ -68,6 +83,99 @@ public OutputLibrary(CodeModel codeModel, BuildContext context)
 
         public IEnumerable<ResponseHeaderGroupType> HeaderModels => (_headerModels ??= EnsureHeaderModels()).Values;
 
+        private static IList<System.Type> GetReferenceClassCollection()
```
Comment (m-nash) [inline src/AutoRest.CSharp/Output/Models/Types/OutputLibrary.cs:86]: Lets extract the following methods into a class called InheritanceChoser

GetReferenceClassCollection
RebuildExactModelInheritance
GetExactMatch
IsEqual
IsAssignable
RebuildModelInheritance

**Review by m-nash — CHANGES_REQUESTED** [review]

File: src/AutoRest.CSharp/Output/Models/Types/ObjectType.cs [hunk 2]
```diff
@@ -111,6 +114,12 @@ public ObjectType(ObjectSchema objectSchema, BuildContext context, bool isResour
         public ObjectTypeConstructor InitializationConstructor => _initializationConstructor ??= BuildInitializationConstructor();
         public ObjectTypeConstructor SerializationConstructor => _serializationConstructor ??= BuildSerializationConstructor();
 
+        public void OverrideInherits(CSharpType cSharpType)
```
Comment (pakrym) [inline src/AutoRest.CSharp/Output/Models/Types/ObjectType.cs:117]: Please don't make models returned from the OutputLibrary mutable. They interact in many ways and values can already be observed before we override them.

Comment (m-nash) [reply src/AutoRest.CSharp/Output/Models/Types/ObjectType.cs:117]: Added ADO item to track this after the merge into feature/v3
https://dev.azure.com/azure-mgmt-ex/DotNET%20Management%20SDK/_workitems/edit/5724

**Review by m-nash — APPROVED** [review]
@chamons @pakrym this is ready to go in if you guys have a chance to take a look

**Pull request merged**